test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml

```diff
@@ -0,0 +1,401 @@
+open Stdune
+open Fiber.O
+module Scheduler = Dune_engine.Scheduler
+module Dune_rpc = Dune_rpc_private
+module Client = Dune_rpc_impl.Client
+module Session = Csexp_rpc.Session
+module Config = Dune_util.Config
+
+let dune_prog =
+  lazy
+    (let path = Env.path Env.initial in
+     Bin.which ~path "dune" |> Option.value_exn |> Path.to_absolute_filename)
+
+let init_chan ~root_dir =
+  let argv = [ "dune"; "rpc"; "init"; "--wait"; "--root"; root_dir ] in
+  let stdout_i, stdout_w = Unix.pipe ~cloexec:true () in
+  let stdin_i, stdin_w = Unix.pipe ~cloexec:true () in
+  let pid =
+    Spawn.spawn ~prog:(Lazy.force dune_prog) ~argv ~stdout:stdout_w
+      ~stdin:stdin_i ()
+    |> Pid.of_int
+  in
+  (pid, Unix.in_channel_of_descr stdout_i, Unix.out_channel_of_descr stdin_w)
+
+let run_client ?handler f =
+  let pid, in_, out = init_chan ~root_dir:"." in
+  let close =
+    lazy
+      (close_out_noerr out;
+       close_in_noerr in_)
+  in
+  Fiber.fork_and_join_unit
+    (fun () ->
+      let+ _ = Scheduler.wait_for_process pid in
+      Lazy.force close)
+    (fun () ->
+      let* chan = Session.create ~socket:false in_ out in
+      let initialize =
+        let id = Dune_rpc.Id.make (Atom "test") in
+        Dune_rpc.Initialize.Request.create ~id
+      in
+      let+ res = Client.connect ?handler chan initialize ~f in
+      Lazy.force close;
+      res)
+
+let read_lines in_ =
+  let* reader = Scheduler.Worker.create () in
+  let in_ = Unix.in_channel_of_descr in_ in
+  let rec loop acc =
+    let* res = Scheduler.Worker.task reader ~f:(fun () -> input_line in_) in
+    match res with
+    | Ok a -> loop (a :: acc)
+    | Error `Stopped -> assert false
+    | Error (`Exn _) -> Fiber.return (String.concat (List.rev acc) ~sep:"\n")
```

Comment:
Print out the exn? (not sure if it's reachable or what, but hiding errors generally makes things harder to debug)

Reply:
Done.